docs(tem): tem x serverless tutorial (#1559)
scaleway/docs-content: tutorials/send-emails-with-serverless-and-tem/index.mdx

---
meta:
  title: Sending emails with Transactional Email and Serverless using Terraform
  description: Discover how to send emails with Transactional Email and Serverless using Terraform
content:
  h1: Sending emails with Transactional Email and Serverless using Terraform
  paragraph: Discover how to send emails with Transactional Email and Serverless using Terraform
tags: transactional-email tem serverless emails terraform
categories:
  - managed-services
  - serverless
dates:
  validation: 2023-04-17
  posted: 2023-05-01
---

In this tutorial, you will learn how to send emails with [Transactional Email](/managed-services/transactional-email/quickstart/) and [Serverless Functions](/serverless/functions/quickstart/) using a Terraform configuration file that builds your entire infrastructure such as a Serverless Function and a website with a registration page hosted on an Object Storage bucket. It also allows you to configure SPF, DKIM and MX records on your sending domain.

<Macro id="iam-requirements" />
 
<Message type="requirement">

>   - You have an account and are logged into the [Scaleway console.](https://console.scaleway.com)

Reviewer: ```suggestion
  - You have an account and are logged into the [Scaleway console](https://console.scaleway.com)
```

  - You have a domain name

>   - You have [downloaded and installed Terraform](https://developer.hashicorp.com/terraform/downloads)

Reviewer: ```suggestion
  - You have [installed Terraform](https://developer.hashicorp.com/terraform/tutorials/aws-get-started/install-cli)
```

</Message>

## Clone the GitHub repository

> We have created a [GitHub repository](https://github.com/quantumsheep/scw-demo-tem) that contains all the code you need to deploy your infrastructure. Open a terminal and clone the repository to your local machine using the following command:

Reviewer: ```suggestion
We have created a [GitHub repository](https://github.com/scaleway/tutorial-tem-serverless-terraform) that contains all the code you need to deploy your infrastructure. Open a terminal and clone the repository to your local machine using the following command:
```

```shell

> git clone https://github.com/quantumsheep/scw-demo-tem

Reviewer: ```suggestion
git clone https://github.com/scaleway/tutorial-tem-serverless-terraform
```
I ported the repository into our official org

```

### Files and directories

> 

Reviewer 2: I think the tuto is missing numbered steps?

> In the `scw-tem-demo` directory, you will find three folders: `infra`, `signup-handler` and `super-product`.

Reviewer: ```suggestion
In the `tutorial-tem-serverless-terraform` directory, you will find three folders: `infra`, `signup-handler` and `super-product`.
```

- The `infra` folder holds `main.tf`, the Terraform configuration file used to deploy your infrastructure.
- The `signup-handler` folder contains the files that will configure your function. The `main.go` file is the entry point for your function.
- The `super-product` folder holds an `index.html.tftpl` file which is a Terraform template. This file contains an HTML template for the registration page of the website we are creating. This template allows you to automatically create and configure the SPF, MX and DKIM records needed for your domain, as well as a Serverless Function.

<Message type="tip">
  If you want to use a specific function, you can replace `${DISPATCHER_ENDPOINT}` (line 74) in the `index.html.tftpl` file with your function's endpoint.
</Message>

## Build and deploy your infrastructure

> Open a terminal and go to the `scw-tem-demo` directory using the following command:

Reviewer: ```suggestion
Open a terminal and go to the `tutorial-tem-serverless-terraform` directory using the following command:
```

```shell

> cd scw-tem-demo/infra

Reviewer: ```suggestion
cd tutorial-tem-serverless-terraform/infra
```

```

Run the following command to download the dependencies and initialize Terraform:

`terraform init`

The following output displays:

```shell
Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/local...
- Finding latest version of hashicorp/archive...
- Finding latest version of scaleway/scaleway...
- Installing hashicorp/local v2.4.0...
- Installed hashicorp/local v2.4.0 (signed by HashiCorp)
- Installing hashicorp/archive v2.3.0...
- Installed hashicorp/archive v2.3.0 (signed by HashiCorp)
- Installing scaleway/scaleway v2.16.3...
- Installed scaleway/scaleway v2.16.3 (signed by a HashiCorp partner, key ID F5BF26CADF6F9614)

Partner and community providers are signed by their developers.

> If you'd like to know more about provider signing, you can read about it here:

Reviewer 2: ```suggestion
<Message type="tip">
If you'd like to know more about provider signing, you can read about it in the [official Terraform documentation](https://www.terraform.io/docs/cli/plugins/signing.html).
</Message>
```

Reviewer 3: This is part of the output, not sure it'll be clear if I add a message box

> https://www.terraform.io/docs/cli/plugins/signing.html

Reviewer 2: ```suggestion
```

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when

> you run "terraform init" in the future.

Reviewer 2: ```suggestion
you run `terraform init` in the future.
```

Reviewer 3: That's part of the output

Terraform has been successfully initialized!

> You may now begin working with Terraform. Try running "terraform plan" to see

Reviewer 2: ```suggestion
You may now begin working with Terraform. Try running `terraform plan` to see
```

Reviewer 3: This is part of the output, not sure I can edit it

any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

Run the following command to deploy your infrastructure:

```shell

> 

Reviewer: ```suggestion
```

Changes to Outputs:
  + handler_url = (known after apply)
  + url         = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.
```

Type "yes", then **Enter**. Terraform will generate your resources.

The following output displays once your resources are generated:

```shell
Apply complete! Resources: 14 added, 0 changed, 0 destroyed.

Outputs:

handler_url = "https://temdemodct73sjr-tem-demo-signup-handler.functions.fnc.fr-par.scw.cloud"
url = "https://tem-demo-super-product.s3-website.fr-par.scw.cloud"
```

## Sending emails

Click the link that displays next to the `url` output. You are redirected to your newly-created website.

Enter your username in the **Username** field.
Enter your recipient's email address in the **Email address** field. For the purposes of this tutorial, we recommend you enter your own email address.
<Lightbox src="scaleway-sign-in-super-website.webp" alt="" />

Click **Sign in**.
Log in to your email address. Your email titled **Welcome to my cool product!** displays in your mailbox.
<Lightbox src="scaleway-tem-email.webp" alt=""/>

> 
> 
> 
> 
> 
> 
> 

Reviewer: ```suggestion
```
